File: v2/src/gid/mutgid.ts
import { Map } from 'immutable'
import type { Id } from './id'
import { GuidId, idFromJSON } from './id'
import type { Gid } from './gid'
import type { Maybe } from '../maybe'

export class MutGid {
  private data = Map<GuidId, Map<GuidId, Id>>()

  get(entity: Id, label: GuidId): Maybe<Id> {
    if (!(entity instanceof GuidId)) return undefined
    return this.data.get(entity)?.get(label)
  }

  edges(entity: Id): Maybe<Map<GuidId, Id>> {
    if (!(entity instanceof GuidId)) return undefined
    return this.data.get(entity)
  }

  set(entity: GuidId, label: GuidId, value: Id): void {
    const edges = this.data.get(entity) ?? Map<GuidId, Id>()
    this.data = this.data.set(entity, edges.set(label, value))
  }

  delete(entity: GuidId, label: GuidId): void {
    const edges = this.data.get(entity)
    if (edges) {
      const newEdges = edges.delete(label)
      this.data = newEdges.size === 0
        ? this.data.delete(entity)
        : this.data.set(entity, newEdges)
    }
  }

  has(entity: GuidId): boolean { return this.data.has(entity) }
  asGid(): Gid { return entity => this.edges(entity) }

  entities(): Iterable<GuidId> {
    return this.data.keys()
  }

  toJSON(): Record<string, Record<string, object>> {
    const result: Record<string, Record<string, object>> = {}
    for (const [entity, edges] of this.data) {
      const edgeObj: Record<string, object> = {}
      for (const [label, value] of edges) {
        edgeObj[label.guid] = value.toJSON()
      }
      result[entity.guid] = edgeObj
    }
    return result
  }

  static fromJSON(json: Record<string, Record<string, unknown>>): MutGid {
    const gid = new MutGid()
    for (const [entityGuid, edges] of Object.entries(json)) {
      const entity = new GuidId(entityGuid)
      for (const [labelGuid, value] of Object.entries(edges)) {
        const label = new GuidId(labelGuid)
        const id = idFromJSON(value)
        if (id) {
          gid.set(entity, label, id)
        }
      }
    }
    return gid
  }
}

File: v2-rust/src/mutgid.rs
use crate::id::{GuidId, Id};
use im::HashMap;
use serde::{Deserialize, Serialize};
use std::collections::HashMap as StdHashMap;

#[derive(Debug, Clone)]
pub struct MutGid {
    data: HashMap<GuidId, HashMap<GuidId, Id>>,
}

impl MutGid {
    pub fn new() -> Self {
        Self {
            data: HashMap::new(),
        }
    }

    pub fn get(&self, entity: &GuidId, label: &GuidId) -> Option<&Id> {
        self.data.get(entity)?.get(label)
    }

    pub fn edges(&self, entity: &GuidId) -> Option<&HashMap<GuidId, Id>> {
        self.data.get(entity)
    }

    pub fn set(&mut self, entity: GuidId, label: GuidId, value: Id) {
        let edges = match self.data.get(&entity) {
            Some(e) => e.update(label, value),
            None => HashMap::unit(label, value),
        };
        self.data.insert(entity, edges);
    }

    pub fn delete(&mut self, entity: &GuidId, label: &GuidId) {
        if let Some(edges) = self.data.get(entity) {
            let new_edges = edges.without(label);
            if new_edges.is_empty() {
                self.data = self.data.without(entity);
            } else {
                self.data.insert(entity.clone(), new_edges);
            }
        }
    }

    pub fn has(&self, entity: &GuidId) -> bool {
        self.data.contains_key(entity)
    }

    pub fn entities(&self) -> impl Iterator<Item = &GuidId> {
        self.data.keys()
    }

    pub fn to_json(&self) -> StdHashMap<String, StdHashMap<String, serde_json::Value>> {
        let mut result = StdHashMap::new();
        for (entity, edges) in self.data.iter() {
            let mut edge_obj = StdHashMap::new();
            for (label, value) in edges.iter() {
                if let Ok(json) = serde_json::to_value(value) {
                    edge_obj.insert(label.guid.clone(), json);
                }
            }
            result.insert(entity.guid.clone(), edge_obj);
        }
        result
    }

    pub fn from_json(
        json: StdHashMap<String, StdHashMap<String, serde_json::Value>>,
    ) -> Result<Self, serde_json::Error> {
        let mut gid = MutGid::new();
        for (entity_guid, edges) in json {
            let entity = GuidId::new(entity_guid);
            for (label_guid, value_json) in edges {
                let label = GuidId::new(label_guid);
                let id: Id = serde_json::from_value(value_json)?;
                gid.set(entity.clone(), label, id);
            }
        }
        Ok(gid)
    }
}

impl Default for MutGid {
    fn default() -> Self {
        Self::new()
    }
}
